### PR TITLE
Render URL's as HTML links in findings

### LINKED ISSUE #3053: Render URL's as HTML links in findings
There are three areas in the Findings, where URL's would help for navigation:

- ``file_path``, used for Sink Filepath and Location when viewing a Finding. When ``source_code_management_uri`` in the Engagement is set properly, the combination of both can be used to generate an URL. This URL would lead the user to view the file in the source code management system.
- ``sast_source_file_path``, used for Source Filepath when viewing a Finding. As for ``file_path``, when ``source_code_management_uri`` in the Engagement is set properly, the combination of both can be used to generate an URL.
- ``references`` often contain one or more URL's for further information to be found on the internet. Although it's already possible to format URL's as Markdown, many importers don't format the URL's like this. So currently you have to use copy and paste to view the content of the links.

I will provide a pull request soon.

### PR DESCRIPTION
**Pull request by StefanFl**
Resolves  #3053

### CODE REVIEW

**Comment by valentijnscholten** [comment]
I get the idea, but I think in the current form it allows XSS via the source file path values in imports.

**Comment by StefanFl** [comment]
Oh yes. Links are now getting bleached, that should avoid the XSS vulnerability. I've made 2 other small changes:
- The markdown_renderer supports the ``target`` attribute now. So we can have links opening in a new tab. 
- The SSLyze JSON parser doesn't format the link as Markdown anymore. This is an excellent manual test case with a complex URL

File: dojo/unittests/test_finding_model.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        self.assertEqual('<a href=\"URL/FilePath\" target=\"_blank\" title=\"FilePath\">FilePath</a>', finding.get_file_path_with_link())
+
+    def test_get_file_path_with_xss_attack(self):
+        test = Test()
+        engagement = Engagement()
+        test.engagement = engagement
+        finding = Finding()
+        finding.test = test
+        finding.file_path = '<SCRIPT SRC=http://xss.rocks/xss.js></SCRIPT>'
+        engagement.source_code_management_uri = '<IMG SRC=javascript:alert(\'XSS\')>'
```
[the remark this answers is not in the document]
Comment (StefanFl) [reply dojo/unittests/test_finding_model.py:60]: It's a test case, it will never show up in the UI

**Review by valentijnscholten — APPROVED** [review]

**Comment by devGregA** [comment]
Just want to chime in that using `safe` has caused issues in the past. Unless I'm misunderstanding. I don't think the benefits out-weigh the risk.

**Comment by StefanFl** [comment]
Using ``safe`` means that the content will not be escaped. This might allow XSS attacks, as mentioned by @valentijnscholten Therefore the content in the three instances is now treated with Mozilla's ``bleach`` in the function ``create_link`` in ``dojo/utils.py`` . bleach will escape everything but the tag ``a`` and the attributes ``href``, ``target`` and ``title``. This should make us as safe as before when showing the links.

**Review by Maffooch — APPROVED** [review]
Explanation is good for me. @devGregA, will leave it up to you to merge.

**Comment by devGregA** [comment]
Calling in reinforcements since I believe I merged the `safe` ones last time that causes issues @madchap @mtesauro @aaronweaver would you please weigh in?

**Comment by devGregA** [comment]
Just another thought, if someone were to get access to the admin interface, I believe it would bypass bleach?

**Comment by devGregA** [comment]
Just to be clear I'm not against this feature, just trying to make sure we've thought through everything.

**Comment by devGregA** [comment]
Answered my own question, it would not because Stefan is using the bleach method via the template, which is definitely a smart way to do it.

**Comment by devGregA** [comment]
One more thought, it may make it easier to call the wrong method in the future, or miss that an unsanitized method was called.

**Comment by StefanFl** [comment]
@devGregA I understand your concerns. We could do something similar to the ``markdown_render`` template tag used for the Reference field and write a template tag, This template tag, let's call it ``safe_link`` would do the bleaching. With this we move control for a safe user interface from the model to the template itself, which makes us more save for the future.

**Comment by devGregA** [comment]
A template tag would definitely help make it more clear. Let me get another opinion though @StefanFl just so you don't have to waste any effort.

**Comment by devGregA** [comment]
I've pinged both Matt and Aaron via email

**Comment by devGregA** [comment]
One other thought on a display tag, display tags (I believe) would require a transform every time. So the bleach option may be the best option. Thinking through ways to try to ensure that unsafe variables don't get called that is also performant...

**Comment by StefanFl** [comment]
@devEricA Would moving the bleach function from the utils to the display tag really make a difference? This should be a fast and cheap function and I think it's called the same amount of time than before.

**Review by mtesauro — APPROVED** [review]
Approved

**Comment by StefanFl** [comment]
@valentijnscholten @devGregA How do we continue? Leave it like this and merge it or move the bleaching in the display tags? Would be great to get this functionality merged.

**Comment by valentijnscholten** [comment]
I was waiting for the tests to complete after I resolved some conflicts.

**Pull request merged**